python/pyarrow/_dataset.pyx

```diff
@@ -1071,21 +1087,57 @@ cdef class Expression:
         return self.wrapped
 
     def equals(self, Expression other):
-        return self.expression.Equals(other.unwrap())
+        return self.expr.Equals(other.unwrap())
 
     def __str__(self):
-        return frombytes(self.expression.ToString())
+        return frombytes(self.expr.ToString())
 
     def validate(self, Schema schema not None):
         cdef:
             shared_ptr[CSchema] sp_schema
             CResult[shared_ptr[CDataType]] result
         sp_schema = pyarrow_unwrap_schema(schema)
-        result = self.expression.Validate(deref(sp_schema))
+        result = self.expr.Validate(deref(sp_schema))
         return pyarrow_wrap_data_type(GetResultValue(result))
 
     def assume(self, Expression given):
-        return Expression.wrap(self.expression.Assume(given.unwrap()))
+        return Expression.wrap(self.expr.Assume(given.unwrap()))
+
+    def __invert__(self):
+        return NotExpression(self)
+
+    def __richcmp__(self, other, int op):
+        operator_mapping = {
+            Py_EQ: CompareOperator.Equal,
+            Py_NE: CompareOperator.NotEqual,
+            Py_GT: CompareOperator.Greater,
+            Py_GE: CompareOperator.GreaterEqual,
+            Py_LT: CompareOperator.Less,
+            Py_LE: CompareOperator.LessEqual
+        }
+
+        if not isinstance(other, Expression):
+            try:
+                other = ScalarExpression(other)
+            except TypeError:
+                return NotImplemented
+
+        return ComparisonExpression(operator_mapping[op], self, other)
+
+    def __and__(self, other):
+        return _binop(AndExpression, self, other)
+
+    def __or__(self, other):
+        return _binop(OrExpression, self, other)
+
+    def is_valid(self):
+        return IsValidExpression(self)
+
+    def cast(self, type, bint safe=True):
+        return CastExpression(self, to=ensure_type(type), safe=safe)
+
+    def isin(self, values):
+        return InExpression(self, pa.array(values))
```

Comment:
Should this (and `cast` above) only be defined on a FieldExpression?

Reply:
It makes sense on scalars, like `ds.scalar(1.0).cast('int64').isin([1, 2, 3])`
But indeed, in the future we'll need to restrict certain operations on specific expression types, like in ibis [1](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L1085), [2](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L1436), [3](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L2590), etc.